Edit Site: Update progress bar to determinate

tyxla opened the pull request:
## What?
This PR builds on top of the work done in #53032 and makes the progress bar determinate.

Blocked by #53767 and needs to be rebased once it lands.

## Why?
To provide the end users with more accurate feedback on their site editor loading experience.

## How?
We believe that using a determinate progress bar instead will be a better loading experience for the user, because it provides better feedback on the loading progress.

This PR also contains a new redux metadata selector that we're separately introducing in #53767.

## Testing Instructions
* Open the site editor.
* Observe the loading progress bar and verify it works well.

### Testing Instructions for Keyboard
None

## Screenshots or screencast <!-- if applicable -->


https://github.com/WordPress/gutenberg/assets/8436925/7e086d2d-3be4-4193-9879-a9b8c2758941

tyxla committed:
Edit Site: Experiment with determinate progress bar

mtias commented:
This is looking promising!

jasmussen commented:
Feels great!

Some questions about how it's measuring progress, if you look closely at this GIF of the progressbar, you'll see that it starts not from zero but from a ways in, and it sits there for just about a second before the page itself fades in:

![after](https://github.com/WordPress/gutenberg/assets/1204802/9b613862-a15b-44ae-a5b7-4b0397cca444)

Give or take a little wiggleroom, it would seem the most user intuitive if the progressbar started from zero, and only milliseconds after completing the track faded the page in. But I guess as far as the loading progress it also counts _loading itself_? And what might explain the second at the end where it suggests loading to be complete but still sits there a bit?

Detail territory, to be clear! But I'm curious. Nice work.

tyxla commented:
> Some questions about how it's measuring progress, if you look closely at this GIF of the progressbar, you'll see that it starts not from zero but from a ways in, and it sits there for just about a second before the page itself fades in:

> Give or take a little wiggleroom, it would seem the most user intuitive if the progressbar started from zero, and only milliseconds after completing the track faded the page in. But I guess as far as the loading progress it also counts loading itself? And what might explain the second at the end where it suggests loading to be complete but still sits there a bit?

Thanks for taking a look so quickly @jasmussen!

Ideally, the progress bar would progress steadily from 0% to 100% smoothly, at the same pace from the beginning to the end. Unfortunately, we don't live in an ideal world, and the network requests don't take the same amount of time, and they can vary in terms of when they start, how long they take, and when they finish. 

However, the good part about all that is that the progress bar displays the actual, real loading progress, defined by the pending and finished requests. It happens to have a head start sometimes because some of the requests finish really quickly, and others, like the ones that load towards the end, are taking a while longer. And we treat requests as equal, regardless of how quick or complex they are. 

The fact that you're experiencing what you described only demonstrates that the site editor progress bar is actually working great and reflects the actual loading that happens behind the scenes! Let me know if that makes sense.

jasmussen commented:
It does make sense. I wonder if this hack would work, simply add `transition: 0.2ms` to the progressbar itself (if it isn't already), so that it doesn't jump from step to step, but animates from step to step. Yes, it adds an intentional animated lag, but it might make it both smoother and to feel alive some of the time when the longer network requests finish. Would that work?

High level though, this seems good to ship, and these things can all be explored separately. Including, _potentially_, the ability to for the determinate slider to temporarily change to being indeterminate, _and back_, if one of the request takes a bit longer than expected.

mtias commented:
A transition or `motion` component wrap would be good to try.

tyxla commented:
@jasmussen @mtias Adding a transition is a great idea! Since it's a separate improvement to the `ProgressBar` component, so I've suggested it in a separate PR: #53877. Please take a look when you get a chance.

tyxla commented:
In the meantime, I've rebased the PR and it's still waiting for a code review before it gets merged. Noting that it also depends on #53767 which introduces a new `@wordpress/data` redux metadata selector.

jasmussen commented:
Nice!

I noticed today when re-testing, that when loading the site editor, this progressbar fades in at the start. I wonder, if it is determinate ideally it'd start at 0 percent, so essentially the progressbar would come into being from nothing.

Can we/should we remove that initial fade-in? Or just make it _really fast_? Small detail.

tyxla committed:
Reduce initial delay from 1s to 0.2s

tyxla commented:
Just rebased to get the transition effect from https://github.com/WordPress/gutenberg/pull/53877 here.

Also in https://github.com/WordPress/gutenberg/pull/53399/commits/e42847e06f391ff79ad085aa4b5edba7f205b9ad I  reduced the initial delay from 1s to 0.2s, so we still keep a minor fade-in effect but still display the progress bar earlier on. 

I've updated the preview in the PR description.

Let me know what you think.

jasmussen commented:
Happy to try this! Let's get a code review and land it.

ciampo reviewed (approved):
Ignoring the code that actually belongs to #53767, this PR LGTM code-wise 🚀 

Let's land #53767, rebase, and merge this PR too!
